File: typescript/src/client/rpc-transport-plugin.ts
import { ClientError, ServerError } from '../types/errors'
import type { PLATRPCMessage, PLATRPCRequest, PLATRPCResponse } from '../rpc'
import type {
  OpenAPIClientTransportOutcome,
  OpenAPIClientTransportPlugin,
  OpenAPIClientTransportRequest,
} from './transport-plugin'

interface RpcTransportRuntime {
  nextRequestId(prefix: string): string
  stringifyHeaders(headers: Record<string, string | number | boolean | undefined>): Record<string, string>
  parseJson(text: string): unknown
  resolveRpcUrl(): string
  ensureRpcSocket(): Promise<WebSocket>
  sendRpcCancel(id: string): Promise<void>
}

interface RpcConnection {
  socket: WebSocket
  result?: PLATRPCResponse
}

export function createRpcTransportPlugin(runtime: RpcTransportRuntime): OpenAPIClientTransportPlugin<RpcConnection> {
  return {
    name: 'rpc',
    canHandle: ({ transportMode }) => transportMode === 'rpc',
    async connect() {
      return { socket: await runtime.ensureRpcSocket() }
    },
    async sendRequest(connection: RpcConnection, request: OpenAPIClientTransportRequest): Promise<void> {
      const rpcRequest: PLATRPCRequest = {
        jsonrpc: '2.0',
        id: runtime.nextRequestId('rpc'),
        operationId: request.operationId,
        method: request.method,
        path: request.path,
        headers: runtime.stringifyHeaders(request.headers),
        input: request.params,
      }
      connection.result = await new Promise<PLATRPCResponse>((resolve, reject) => {
        const abort = () => {
          void runtime.sendRpcCancel(rpcRequest.id)
          reject(new DOMException('RPC request was aborted', 'AbortError'))
        }
        if (request.signal?.aborted) {
          abort()
          return
        }
        const onMessage = async (event: Event) => {
          if (!(event instanceof MessageEvent)) return
          const payload = runtime.parseJson(String(event.data)) as PLATRPCMessage
          if (!payload || typeof payload !== 'object' || payload.id !== rpcRequest.id) return
          if ('event' in payload && payload.event) {
            await request.onEvent?.({ id: payload.id, event: payload.event, data: payload.data })
            return
          }
          connection.socket.removeEventListener('message', onMessage)
          request.signal?.removeEventListener('abort', abort)
          resolve(payload as PLATRPCResponse)
        }
        connection.socket.addEventListener('message', onMessage)
        request.signal?.addEventListener('abort', abort, { once: true })
        connection.socket.send(JSON.stringify(rpcRequest))
      })
    },
    async getResult(connection: RpcConnection, request: OpenAPIClientTransportRequest): Promise<OpenAPIClientTransportOutcome> {
      const response = connection.result!
      if (response.ok) return { id: request.id, ok: true, result: response.result }
      const status = response.error.status ?? 500
      const common = {
        url: runtime.resolveRpcUrl(),
        method: request.method,
        status,
        statusText: response.error.message,
        headers: new Headers(),
        bodyText: JSON.stringify(response.error.data ?? response.error.message),
        bodyJson: response.error.data ?? response.error.message,
      }
      return {
        id: request.id,
        ok: false,
        error: status >= 400 && status < 500 ? new ClientError(common) : new ServerError(common),
      }
    },
  }
}

File: python/z2/rpc_transport_plugin.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Protocol

from .transport_plugin import OpenAPIClientTransportPlugin, TransportRequest, TransportResult


class _RPCRuntime(Protocol):
    def next_rpc_id(self) -> str:
        ...

    def resolve_rpc_url(self) -> str:
        ...

    def _coerce_input_model(self, input_data: Any) -> Any:
        ...


@dataclass
class _RPCConnection:
    operation: Any
    response: Any = None


def create_rpc_transport_plugin(runtime: _RPCRuntime, operation: Any, *, is_async: bool = False) -> OpenAPIClientTransportPlugin:
    class RPCTransportPlugin:
        name = "rpc"

        def can_handle(self, request: dict[str, str]) -> bool:
            return request.get("transport_mode") == "rpc"

        def connect(self, request: TransportRequest) -> _RPCConnection:
            return _RPCConnection(operation=operation)

        def send_request(self, connection: _RPCConnection, request: TransportRequest) -> None:
            connection.response = None

        def get_result(self, connection: _RPCConnection, request: TransportRequest) -> Any:
            request_payload = {
                "jsonrpc": "2.0",
                "id": runtime.next_rpc_id(),
                "operationId": connection.operation.operation_id,
                "method": connection.operation.method,
                "path": connection.operation.path,
                "headers": request.headers,
                "input": runtime._coerce_input_model(request.params),
            }
            from websockets.sync.client import connect
            with connect(runtime.resolve_rpc_url()) as websocket:
                websocket.send(json.dumps(request_payload))
                while True:
                    response = json.loads(websocket.recv())
                    if response.get("event"):
                        if request.on_event is not None:
                            maybe = request.on_event(response)
                            if is_async and hasattr(maybe, "__await__"):
                                import asyncio
                                asyncio.run(maybe)
                        continue
                    break
            if not response.get("ok"):
                error = response.get("error") or {}
                return TransportResult(id=request.id, ok=False, error=RuntimeError(error.get("message", "RPC request failed")))
            return TransportResult(id=request.id, ok=True, result=response.get("result"))

    return RPCTransportPlugin()
